### ISSUE TITLE
Repeated ID linter gives false positive on neighbouring conditional branches

### ISSUE THREAD

**Issue by daveallie** [sds/haml-lint#219]
### Input File:

```haml
- if something?
  #my-id Some text 1
  #my-id2 Some text 2
  #my-id3 Some text 3
- else
  #my-id Some other text 1
  #my-id2 Some other text 2
  #my-id3 Some other text 3
```

---

### Reports:

scratch.haml:2 [E] RepeatedId: Do not repeat id "#my-id" on the page
scratch.haml:6 [E] RepeatedId: Do not repeat id "#my-id" on the page
scratch.haml:3 [E] RepeatedId: Do not repeat id "#my-id2" on the page
scratch.haml:7 [E] RepeatedId: Do not repeat id "#my-id2" on the page
scratch.haml:4 [E] RepeatedId: Do not repeat id "#my-id3" on the page
scratch.haml:8 [E] RepeatedId: Do not repeat id "#my-id3" on the page

---

### Should Report:

Nothing

---

Not sure if it's a possibility, but it'd be really nice if haml_lint could detect conditional ruby branches and only run the Repeated ID selector on accessible elements through each pathway.

**Comment by trotzig** [comment]
Ah, good catch! I wonder if this is worth fixing, perhaps it's better to just locally ignore the RepeatedId linter? Let's see if @michaelherold has any good input as well.

**Comment by daveallie** [comment]
Yeah, at the moment I am just ignoring the whole file as it is coming up in a switch block with a bunch of cases. It's pretty tedious to go through each case and put disable/enable blocks around elements with repeated ids, especially if there are elements between the ones with repeated ids.

**Comment by michaelherold** [comment]
Thanks for the reproduction steps! I'll try to figure out how we can scope that in the AST.

**Comment by andyw8** [comment]
I don't think it's possible to fully handle this with static analysis.

For example, say the markup was:

```haml
- if 1 < 2
  #my-id Something
- elsif 1 > 0
  #my-id Something
- else
  #my-other-id Something else
```

The linter would need to determine which branches of the conditional would evaluate to true.

I think this would be better handled by something which ran after the view was rendered, e.g.

```ruby
visit 'widgets'
expect(page).to_not have_repeated_ids
```

**Comment by daveallie** [comment]
> The linter would need to determine which branches of the conditional would evaluate to true.

Not necessarily. If the linter could view the possible code paths as a tree, then the path to each leaf needs to have at most one of each id selector. In the example you gave, I wouldn't expect the linter to raise as it considers lines 1 and 3 different code paths that can't be hit at the same time.

I think the bigger issue with this approach is how big the search space gets. e.g. If I had 3 case statements, each with 3 branches, that's 9 possible paths. And if I add an if/else to each of the case branches, that's 36 possible paths.

It's going to blow out very quickly, but maybe I am just over exaggerating the problem.

**Comment by andyw8** [comment]
@daveallie That was a bad example I gave, I was intending something like this:

``` haml
- if 1 < 2
  #my-id Something
- if 1 > 0
  #my-id Other Something
```

The situation where only a single clause can be true is indeed simpler, although as you say, could be slow if there are many possible paths.

I wonder if there are any linters already in RuboCop which follow this kind of pattern.

**Comment by michaelherold** [comment]
🤔 I've been looking into this one and I'm stumped.

This simple Haml file:

```haml
- if 1 < 2
  #my-id Hello
- else
  #my-id World
```

Gets parsed into this structure:

```
                         RootNode
                            |
                SilentScriptNode(if i < 2)
               /            |              \
TagNode(Hello)    SilentScriptNode(else)    TagNode(World)
```

This isn't what I expected. I expected something like this:

```
                            RootNode
                           /        \
 SilentScriptNode(if i < 2)          SilentScriptNode(else)
           |                                   |
     TagNode(Hello)                      TagNode(World)
```

The latter would make sense to parse through: you would only have to look at the parent node's ancestors and prior siblings to check whether an ID has been used prior to the current node.

However, since the Haml is parsed into the former structure, I'm not really clear on how to tell which nodes you would have to look at for a duplicate. The rendered output will only have the `TagNode(Hello)` tag, so there isn't a duplicate ID ... but it's not clear how you would divine this from the parsed tree structure.

If anyone has any ideas on how we can handle this, I would love to hear them!

**Comment by trotzig** [comment]
Wow, that's confusing...
